django/applications/catmaid/static/js/widgets/tagtable.js

```diff
@@ -0,0 +1,369 @@
+/* -*- mode: espresso; espresso-indent-level: 8; indent-tabs-mode: t -*- */
+/* vim: set softtabstop=2 shiftwidth=2 tabstop=2 expandtab: */
+
+(function(CATMAID) {
+
+  "use strict";
+
+  var CACHE_TIMEOUT = 5*60*1000;  // cache invalidation timeout in ms
+
+  var TagTable = function() {
+    this.widgetID = this.registerInstance();
+    this.selectedSkeletons = new CATMAID.BasicSkeletonSource(this.getName());
+  };
+
+  $.extend(TagTable.prototype, new InstanceRegistry());
+
+  TagTable.prototype.getName = function() {
+    return 'Tag Table ' + this.widgetID;
+  };
+
+  var labelSkelMappingCache = {};  // todo - button for enabling/disabling cache?
+
+  /**
+   *
+   * @param labelIDs - array of label IDs
+   * @param callback - function which takes a list of skelIDs
+   */
+  var processSkelIDsFromLabelIDs = function(labelIDs, callback) {
+    var remoteLabelIDs = [];
+    var skelIDs = new Set();
+
+    var now = Date.now();
+
+    for (var i = 0; i < labelIDs.length; i++) {
+      var labelID = labelIDs[i];
+      if (labelID in labelSkelMappingCache && now - labelSkelMappingCache[labelID].timestamp <= CACHE_TIMEOUT) {
+        skelIDs.addAll(
+          labelSkelMappingCache[labelID].skelIDs
+        );
+      } else {
+        remoteLabelIDs.push(labelID);
+      }
+    }
+
+    if (remoteLabelIDs.length === 0) {
+      callback(Array.from(skelIDs));
+      return;
+    }
+
+    CATMAID.fetch(project.id + '/skeletons/node_labels', 'POST', {
+      'label_ids': remoteLabelIDs
+    }).then(function(json) {
+      now = Date.now();
+
+      for (var i = 0; i < json.length; i++) {
+        var labelSkelTuple = json[i];
+        labelSkelMappingCache[labelSkelTuple[0]].skelIDs = labelSkelTuple[1];
+        labelSkelMappingCache[labelSkelTuple[0]].timestamp = now;
+        skelIDs.addAll(labelSkelTuple[1]);
+      }
+
+      callback(Array.from(skelIDs));  // todo - replace with promise?
```

Comment:
Agreed! Like explained this morning Chris: you would only need to return `Array.from(skelIDs)` from the then-function. And above, if there are no labels, you'd return `Promse.resolve(Array.from(skelIDs));`. The `callback` parameter can be removed, and the calling functions would call without the callback and call `then(<original-callback-function>);` on the returned promise.